### PR TITLE
Move nonce/pow to end of block header, change post-pow hash

### PR DESCRIPTION
**Pull request by yeastplume**
Changes as per https://github.com/mimblewimble/grin/issues/780

 - [x] Block header ordering updated
 - [x] Update call to cuckoo miner async mode
 - [x] Update post-pow hash to blake2(cuckoo_cycle(blake2(pre-pow))),
 - [x] Check/update all tests

### CODE REVIEW

**Comment by yeastplume** [comment]
Tests pass locally, now to refresh Travis until it feels like accepting them.

Ready for review anyhow.. it's always a bit awkward to try to serialise objects in different ways depending on the context, but I think a specific pre_pow hash function for the block header makes sense. 

Also note that since we're essentially using the header pow hash as the block hash, we need a way to populate the pow randomly during tests that don't care about the POW to prevent duplicate blocks. I considered changing the default BH struct to use a random POW, but I don't think that's as clean.

File: core/src/core/block.rs [hunk 1]
```diff
@@ -135,33 +135,21 @@ impl Default for BlockHeader {
...
-			[write_fixed_bytes, &self.kernel_root]
-		);
-
-		try!(writer.write_u64(self.nonce));
-		try!(self.total_difficulty.write(writer));
-		try!(self.total_kernel_offset.write(writer));
-
-		if writer.serialization_mode() != ser::SerializationMode::Hash {
+		if writer.serialization_mode() == ser::SerializationMode::Hash {
 			try!(self.pow.write(writer));
```
Comment (antiochp) [inline core/src/core/block.rs]: From eavesdropping on prior discussions it sounds the `pow` effectively encodes everything we care about - so we can just hash this (which is presumably faster)?

So effectively `block.hash() == header.hash() == pow.hash()`? Which has a nice feel to it.

Comment (yeastplume) [reply core/src/core/block.rs]: Yes ... all seems to slot nicely in place with this approach, and everything commits to exactly as much as it needs to and no more
